README.md
fluentd-standalone [![Build Status](https://travis-ci.org/xerial/fluentd-standalone.svg?branch=develop)](https://travis-ci.org/xerial/fluentd-standalone) ![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.xerial/fluentd-standalone_2.12/badge.svg) [![Scaladoc](http://javadoc-badge.appspot.com/org.xerial/fluentd-standalone_2.12.svg?label=scaladoc)](http://javadoc-badge.appspot.com/org.xerial/fluentd-standalone_2.12)
=========

Standalone fluentd (http://fluentd.org) server for Java and Scala.

This library launches a new fluentd server using a random port when fluentd is not started in the node. 

Use cases:
  * Running test codes that use fluentd (logger) in a build server (e.g., Travis CI, Jenkins, etc.), in which fluentd is not running.
  

## Usage (Java/Scala)

fluentd depends on some ruby modules. You need to install these dependencies first.

The simplest way is just install fluentd.
```
$ gem install fluentd -v 0.14.11
```

See also http://docs.fluentd.org/articles/quickstart.

### Maven
```xml
<dependencies>
  ...
  <dependency>
    <groupId>org.xerial</groupId>
    <artifactId>fluentd-standalone_2.12</artifactId>
    <version>0.14.11</version>
  </dependency>
  ...
</dependencies>
```

### sbt
```
libraryDependencies += "org.xerial" %% "fluentd-standalone" % "0.14.11"
```

## Sample code

```java
import xerial.fluentd.FluentdStandalone;

FluentdStandalone s = new FluentdStandalone();
// Start new fluentd server
s.start();

// Send log messsages to fluentd
int port = s.port(); // fluentd port

// Terminate the fluentd
s.stop();
```


## For developers

```
$ git clone https://github.com/xerial/fluentd-standalone.git
$ cd fluentd-standalone
$ git submodule init          # Only for the first time
$ git submodule update        # Fetch fluentd 

# Run tests
$ ./sbt test 
```


RELEASE_NOTES.md
# Release Notes

## 0.14.11

 - Upgrade to fluentd 0.14.11
 - Scala 2.12, 2.11 support.
 - Dropped Scala 2.10 support

## 0.1.2
 - Fixed temporary folder to extract fluentd

## 0.1
 - The first release


src/test/scala/xerial/fluentd/MySpec.scala
package xerial.fluentd

import org.scalatest._
import xerial.core.io.Resource
import xerial.core.util.Timer
import wvlet.log.Logger
import wvlet.log.LogSupport
import wvlet.log.LogFormatter.SourceCodeLogFormatter
import scala.language.implicitConversions

//--------------------------------------
//
// MySpec.scala
// Since: 2012/11/20 12:57 PM
//
//--------------------------------------

/**
 * Helper trait for writing test codes. Extend this trait in your test classes
 */
trait MySpec extends WordSpec with Matchers with GivenWhenThen with OptionValues with Resource with Timer with LogSupport with BeforeAndAfter {

  Logger.setDefaultFormatter(SourceCodeLogFormatter)

  implicit def toTag(s:String) = Tag(s)

}


src/test/scala/xerial/fluentd/FluentdStandaloneTest.scala
//--------------------------------------
//
// FluentdStandaloneTest.scala
// Since: 2013/12/16 10:50 AM
//
//--------------------------------------

package xerial.fluentd

import java.net.Socket
import xerial.core.util.Shell


/**
 * @author Taro L. Saito
 */
class FluentdStandaloneTest extends MySpec {

  "FluentdStandaloe" should {

    "launch fluentd" in {

      val fluentd = FluentdStandalone.start()

      try {
        val port = fluentd.port
        val catCommand = s"${fluentd.config.fluentCatCmd} -p ${port} debug"
        val ret = Shell.exec("echo '{\"message\":\"hello\"}' | " + catCommand)
        Thread.sleep(1000)
        ret shouldBe (0)
      }
      finally {
        fluentd.stop
      }
    }


  }

}

src/main/scala/xerial/fluentd/FluentdStandalone.scala
//--------------------------------------
//
// FluentdStandalone.scala
// Since: 2013/12/16 10:48 AM
//
//--------------------------------------

package xerial.fluentd

import xerial.core.io.{Resource, IOUtil}
import java.io._

import wvlet.log.LogSupport
import xerial.core.io.Path._
import xerial.core.util.Shell
import xerial.core.io.Resource.VirtualFile
import java.net.Socket

/**
 * @author Taro L. Saito
 */
object FluentdStandalone extends LogSupport {

  def start(config:FluentdConfig = FluentdConfig()) : FluentdStandalone = {
    val fd = new FluentdStandalone(config)
    fd.startAndAwait
    fd
  }

  def defaultConfig : String =
    """
      |## Forward all log messages to stdout
      |<match **>
      |  type stdout
      |</match>
      |
    """.stripMargin

}


/**
 * Fluentd configuration
 * @param port port number to listen
 * @param configFile if null, the default configuration file is generated to (workdir)/fluent.conf
 * @param workDir working directory. default is target/fluentd
 * @param configuration fluentd configuration script used when configFile is null. The default configuration just outputs every log to stdout.
 */
case class FluentdConfig(port:Int = IOUtil.randomPort,
                         configFile:String=null,
                         initialWaitMilliSeconds:Int = 500,
                         maxWaitMilliSeconds:Int = 10000,
                         workDir:String = "target/fluentd",
                         configuration : String = FluentdStandalone.defaultConfig) {


  def getConfigFile : File = if(configFile == null) {
    workDir / "fluent.conf"
  }
  else
    new File(configFile)


  def fluentdCmd = s"${workDir}/core/bin/fluentd"
  def fluentCatCmd = s"${workDir}/core/bin/fluent-cat"

}



class FluentdStandalone(val config:FluentdConfig) extends LogSupport {

  def this(port:Int) = this(FluentdConfig(port = port))

  private var fluentdProcess : Option[Process] = None


  def port : Int = config.port

  /**
   * Start fluentd and returns fluentd port number
   * @return
   */
  def start : Int = {
    prepare(config)

    info(s"Starting fluentd")
    val process = Shell.launchProcess(s"${config.fluentdCmd} -c ${config.getConfigFile} --no-supervisor")
    fluentdProcess = Some(process)
    val t = new Thread(new Runnable with LogSupport {
      def run() {
        process.waitFor()
        val ret = process.exitValue
        ret match {
          case 0 => // OK
          case 143 => // terminated by stop() (SIGTERM)
          case code =>
            error(s"Error occurred while launching fluentd (error code:$code). If you see 'LoadError', install fluentd and its dependencies by 'gem install fluentd'")
        }
      }
    })
    t.setDaemon(true)
    t.start()

    config.port
  }

  /**
   * Start fluentd and await its startup
   * @return fluentd port number
   */
  def startAndAwait : Int = {
    val port = start

    // Wait until fluentd starts
    var waitDuration = config.initialWaitMilliSeconds
    var count = 0
    var connected = false
    while(!connected && waitDuration < config.maxWaitMilliSeconds) {
      Thread.sleep(waitDuration)
      try {
        val sock = new Socket("localhost", port)
        sock.close()
        connected = true
      }
      catch {
        case e:IOException =>
          warn(s"${e.getMessage}")
          waitDuration = (waitDuration * 1.5).toInt
          count += 1
      }
    }

    if(!connected)
      throw new IOException("Failed to connect fluentd")

    port
  }


  def stop {
    fluentdProcess.map { p =>
      info(s"Terminating fluentd")
      p.destroy()
    }
  }


  private[fluentd] def prepare(config:FluentdConfig) = {

    def mkdir(path:File) {
      path.mkdirs()
      if(!path.exists())
        throw new IOException(s"Failed to create directory: ${path}")
    }

    // Create workdir

    val workDir = new File(config.workDir)
    val coreDir = workDir / "core"
    val pluginDir = workDir / "plugin"
    debug(s"Creating fluentd workdir: $workDir")
    mkdir(workDir)
    mkdir(coreDir)
    mkdir(pluginDir)

    // Copy fluentd
    for(f <- Resource.listResources("xerial.fluentd.core")) {

      def relPath(f:VirtualFile) = f.logicalPath.replaceFirst("xerial/fluent/core/", "")
      val rpath = relPath(f)
      val targetPath = coreDir / rpath

      if(f.isDirectory)
        targetPath.mkdirs()
      else {
        IOUtil.withResource(f.url.openStream) { r =>
          IOUtil.readFully(r) { fileContents =>
            targetPath.getParentFile().mkdirs()
            IOUtil.withResource(new FileOutputStream(targetPath)) { out =>
              out.write(fileContents)
            }

            if(rpath.startsWith("bin"))
              targetPath.setExecutable(true)
          }
        }
      }
    }

    def fluentConf(port:Int, configuration:String) =
      s"""
        |## Listen a socket
        |<source>
        |  type forward
        |  port ${port}
        |</source>
        |
        |${configuration}
      """.stripMargin

    // Create fluent.conf if it doesn't specified
    if(config.configFile == null) {
      val targetFile = config.getConfigFile
      // Create fluent.conf from a template
      val configText = fluentConf(config.port, config.configuration)
      // Write fluent.conf
      IOUtil.withResource(new FileWriter(targetFile)) { writer =>
        writer.write(configText)
      }
    }
  }

}


project/Build.scala
import sbt._
import sbt.Keys._

object Build extends sbt.Build {

  private def profile = System.getProperty("xerial.profile", "default")

  def releaseResolver(v: String): Option[Resolver] = {
    profile match {
      case "default" => {
        val nexus = "https://oss.sonatype.org/"
        if (v.trim.endsWith("SNAPSHOT"))
          Some("snapshots" at nexus + "content/repositories/snapshots")
        else
          Some("releases" at nexus + "service/local/staging/deploy/maven2")
      }
      case p => {
        scala.Console.err.println("unknown xerial.profile '%s'".format(p))
        None
      }
    }
  }

  val copyFluentd = TaskKey[Unit]("copy-fluentd", "Embed fluend into jar")

  val SCALA_VERSION = "2.12.1"

  lazy val root = Project(
    id = "fluentd-standalone",
    base = file("."),
    settings = Defaults.defaultSettings ++ sbtrelease.ReleasePlugin.releaseSettings ++
      Seq(
        organization := "org.xerial",
        organizationName := "xerial.org",
        organizationHomepage := Some(new URL("http://github.com/xerial/fluentd-standalone")),
        description := "Standalone fluend server for Java and Scala",
        scalacOptions ++= Seq("-unchecked", "-deprecation", "-feature"),
        // custom settings here
        scalaVersion := SCALA_VERSION,
        crossScalaVersions := Seq("2.10.5", "2.11.8", SCALA_VERSION),
        crossPaths := true,
        publishMavenStyle := true,
        publishArtifact in Test := false,
        publishTo <<= version { v => releaseResolver(v) },
        compile in Compile <<= (compile in Compile).dependsOn(copyFluentd),
        copyFluentd := {
          val baseDir : File = baseDirectory.value
          val fluentd = baseDir / "fluentd"

          Seq("2.10", "2.11", "2.12").foreach { ScalaV =>
            val targetDir : File = target.value / s"scala-$ScalaV" / "classes/xerial/fluentd/core"
            val s = streams.value
            def rpath(path:File) =  path.relativeTo(baseDir).getOrElse(path)

            s.log.info("copy " + rpath(fluentd) + " to " + rpath(targetDir))
            val p = (fluentd ** "*")
            for(file <- p.get; relPath <- file.relativeTo(fluentd) if !relPath.getPath.startsWith(".git")) {
              val out = targetDir / relPath.getPath
              if(file.isDirectory) {
                s.log.debug("create directory: " + rpath(out))
                IO.createDirectory(out)
              }
              else {
                s.log.debug("copy " + rpath(file) + " to " + rpath(out))
                IO.copyFile(file, out, preserveLastModified=true)
              }
            }
          }
        },
        logBuffered in Test := false,
        libraryDependencies ++= Seq(
          // Add dependent jars here
          "org.xerial" %% "xerial-core" % "3.6.0",
          "org.slf4j" % "slf4j-simple" % "1.7.22" % "test",
          "org.scalatest" %% "scalatest" % "3.0.1" % "test"
        ),
        pomExtra := {
          <url>http://xerial.org/</url>
            <licenses>
              <license>
                <name>Apache 2</name>
                <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
              </license>
            </licenses>
            <scm>
              <connection>scm:git:github.com/xerial/fluentd-standalone.git</connection>
              <developerConnection>scm:git:git@github.com:xerial/fluentd-standalone.git</developerConnection>
              <url>github.com/xerial/fluentd-standalone.git</url>
            </scm>
            <properties>
              <scala.version>{SCALA_VERSION}</scala.version>
              <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
            </properties>
            <developers>
              <developer>
                <id>xerial</id>
                <name>Taro L. Saito</name>
                <url>http://xerial.org/leo</url>
              </developer>
            </developers>
        }

      )
  )
}
